README.md
# node-couchdb [![Build Status](https://secure.travis-ci.org/1999/node-couchdb.svg?branch=master)](http://travis-ci.org/1999/node-couchdb) [![Dependency Status](https://david-dm.org/1999/node-couchdb.svg)](https://david-dm.org/1999/node-couchdb) [![devDependency Status](https://david-dm.org/1999/node-couchdb/dev-status.svg)](https://david-dm.org/1999/node-couchdb#info=devDependencies) [![Greenkeeper badge](https://badges.greenkeeper.io/1999/node-couchdb.svg)](https://greenkeeper.io/)

`node-couchdb` package provides an easy way to interact with CouchDB using preferred cache layer:

 * [process memory](https://www.npmjs.com/package/node-couchdb-plugin-memory)
 * [memcached](https://www.npmjs.com/package/node-couchdb-plugin-memcached)
 * place for your plugin :)

# Installation
``` bash
npm install node-couchdb --save
```

# API
## Constructor
`node-couchdb` exports constructor, which accepts one object argument with properties `host` (127.0.0.1 by default), `port` (5984 by default), `protocol` (http by default), `cache` (one of plugins, null by default), `auth` (object with properties `{user, pass}`) and `timeout` for all requests (5000 by default). All object fields are optional.

ES Module:
```javascript
import NodeCouchDb from 'node-couchdb';
```

Common JS:
```javascript
const NodeCouchDb = require('node-couchdb');
```

```javascript
// node-couchdb instance with default options
const couch = new NodeCouchDb({
    auth: {
        user: AUTH_USER,
        pass: AUTH_PASS
    }
});

// node-couchdb instance with Memcached
const MemcacheNode = require('node-couchdb-plugin-memcached');
const couchWithMemcache = new NodeCouchDb({
    cache: new MemcacheNode,
    auth: {
        user: AUTH_USER,
        pass: AUTH_PASS
    }
});

// node-couchdb instance talking to external service
const couchExternal = new NodeCouchDb({
    host: 'couchdb.external.service',
    protocol: 'https',
    port: 6984,
    auth: {
        user: AUTH_USER,
        pass: AUTH_PASS
    }
});

```

All node-couchdb methods return Promise instances which resolve if everything works as expected and reject with Error instance which usually has `code` and `body` fields. See package source and tests for more info.

## Create database
```javascript
couch.createDatabase(dbName).then(() => {...}, err => {
    // request error occured
});
```

## Drop database
```javascript
couch.dropDatabase(dbName).then(() => {...}, err => {
    // request error occured
});
```

## List databases
```javascript
couch.listDatabases().then(dbs => dbs.map(...), err => {
    // request error occured
});
```

## Get document by its id
```javascript
couch.get("databaseName", "some_document_id").then(({data, headers, status}) => {
    // data is json response
    // headers is an object with all response headers
    // status is statusCode number
}, err => {
    // either request error occured
    // ...or err.code=EDOCMISSING if document is missing
    // ...or err.code=EUNKNOWN if statusCode is unexpected
});
```

## Get view results
```javascript
const dbName = "database";
const startKey = ["Ann"];
const endKey = ["George"];
const viewUrl = "_design/list/_view/by_firstname";

const queryOptions = {
    startkey: startKey,
    endkey: endKey
};

couch.get(dbName, viewUrl, queryOptions).then(({data, headers, status}) => {
    // data is json response
    // headers is an object with all response headers
    // status is statusCode number
}, err => {
    // either request error occured
    // ...or err.code=EDOCMISSING if document is missing
    // ...or err.code=EUNKNOWN if statusCode is unexpected
});
```

## Query using Mango
```javascript
const dbName = "database";
const mangoQuery = {
    selector: {
        firstname: {
            $gte: 'Ann',
            $lt: 'George'
        }
    }
};

couch.mango(dbName, mangoQuery).then(({data, headers, status}) => {
    // data is json response
    // headers is an object with all response headers
    // status is statusCode number
}, err => {
    // either request error occured
    // ...or err.code=EDOCMISSING if document is missing
    // ...or err.code=EUNKNOWN if statusCode is unexpected
});
```

## Insert a document
```javascript
couch.insert("databaseName", {
    _id: "document_id",
    field: ["sample", "data", true]
}).then(({data, headers, status}) => {
    // data is json response
    // headers is an object with all response headers
    // status is statusCode number
}, err => {
    // either request error occured
    // ...or err.code=EDOCCONFLICT if document with the same id already exists
});
```

## Update a document
```javascript
// note that "doc" must have both "_id" and "_rev" fields
couch.update("databaseName", {
    _id: "document_id",
    _rev: "1-xxx"
    field: "new sample data",
    field2: 1
}).then(({data, headers, status}) => {
    // data is json response
    // headers is an object with all response headers
    // status is statusCode number
}, err => {
    // either request error occured
    // ...or err.code=EFIELDMISSING if either _id or _rev fields are missing
});
```

## Insert an attachment
```javascript
couch.insertAttachment("databaseName", "document id", "attachment name", "attachment body", "doc revision").then(({data, headers, status}) => {
    // data is json response
    // headers is an object with all response headers
    // status is statusCode number
}, err => {
    // either request error occured
    // ...or err.code=EFIELDMISSING if either _id or _rev fields are missing
});
```

## Delete an attachment
```javascript
// note that "doc" must have both "_id" and "_rev" fields
couch.update("databaseName", "document id", "attachment name", "doc revision").then(({data, headers, status}) => {
    // data is json response
    // headers is an object with all response headers
    // status is statusCode number
}, err => {
    // either request error occured
    // ...or err.code=EFIELDMISSING if either _id or _rev fields are missing
});
```

## Use an update function
```javascript
couch.updateFunction("databaseName", "designDocument", "updateFunction", {optional query string}, "docid").then(({data, headers, status}) => {
    // data is json response
    // headers is an object with all response headers
    // status is statusCode number
}, err => {
    // either request error occured
    // ...or err.code=EFIELDMISSING if either _id or _rev fields are missing
});
```

## Delete a document
```javascript
couch.del("databaseName", "some_document_id", "document_revision").then(({data, headers, status}) => {
    // data is json response
    // headers is an object with all response headers
    // status is statusCode number
}, err => {
    // either request error occured
    // ...or err.code=EDOCMISSING if document does not exist
    // ...or err.code=EUNKNOWN if response status code is unexpected
});
```

## Generate unique identifier(s)
```javascript
// get one unique id
couch.uniqid().then(ids => ids[0]);

// get N unique ids
couch.uniqid(N).then(ids => ids.map(...));
```


HISTORY.md
# 2.1.0

 * **node version support**: Removes legacy Node support. Only Node >= 18 supported.
 * **new**: Removes the node-fetch package, since the Fetch API is supported natively in Node 18 (and in [Bun](https://bun.sh))

# 2.0.0

 * **breaking change**: Removes legacy Node support. Only Node >= 14 supported.
 * **new**: Replaces deprecated request package with node-fetch, which will allow a smoother transition to global fetch when it is no longer experimental.
 * **new**: Updates dev dependencies including Babel for a clean npm audit
 * **new**: Changes tests to remove assumption of admin party. By default, CouchDB v3 (the current supported version) requires an admin user
 * **new**: Adds a test for the mango method, and removes the "query" argument from it (seems like there isn't a need for it). 
 * **fix**: Corrects the example on the README for the mango method.
 * **fix**: Fixes bug in test "should return promise for createDatabase operation" where the created database was not dropped afterward

# 1.3.0

 * **new**: Support for [update functions](http://docs.couchdb.org/en/2.0.0/couchapp/ddocs.html#update-functions) ([@Flamenco](https://github.com/Flamenco))
 * **new**: Support for inserting and deleting attachments ([@bgardner87](https://github.com/bgardner87))

# 1.2.0

 * **new**: New `mango` API support ([@Flamenco](https://github.com/1999/node-couchdb/pull/20))

# 1.1.1
 * **fix**: CouchDB 400 status codes now produce `EBADREQUEST` error code ([@pvomhoff](https://github.com/pvomhoff))
 * **fix**: `UnhandledPromiseRejectionWarning` inside tests is fixed ([@pvomhoff](https://github.com/pvomhoff))

# 1.1.0

 * **new**: you can pass new `auth` field into node-couchdb constructor argument to support non-admin-party CouchDB mode. See [README](https://github.com/1999/node-couchdb#constructor) for more info.

# 1.0.x

There had been problems with extending and support of `node-couchdb` since its first release due to its code. Tests were a bit messy and code was too old. Also there was an architecture problem with cache layers: internally they were acting really weird. `node-couchdb@1` was designed to be extremely simple to read and extend. Cache layers were renamed into plugins and separated from the package itself. Tests were rewritten in `mocha`. Library code was rewritten in ES2015.

 * **breaking change**: `node-couchdb` is now ES2015-compatible: only node.js >= 4 is supported. If you're using node.js 0.10 or 0.12, `require('node-couchdb/dist/legacy')`, but it's not tested. You have been warned!
 * **breaking change**: new API: package exports constructor with new arguments, all methods return promise instance
 * **breaking change**: cache layers have been removed from the package into separate NPM packages: [memcached](https://www.npmjs.com/package/node-couchdb-plugin-memcached), [process memory](https://www.npmjs.com/package/node-couchdb-plugin-memory). If you're missing smth feel free to add a new one and publish it with `node-couchdb-plugin` keyword. Also send a PR to `node-couchdb` README about your plugin and it will be added to the list.
 * **new**: promises can be rejected with Error instance which usually has `code` and `body` fields. Look tests and README for more examples
 * **new**: all tests rewritten in `mocha` + many new tests introduced

Note about writing your own node-couchdb-plugin: grab tests from `node-couchdb-plugin-memory` and enhance with your own. Package should export constructor and the prototype should have only three methods (set, get, invalidate), so it should be pretty simple to write tests for it.

# 0.5.0

 * new: createDatabase invokes callback with EDBEXISTS error if database already exists
 * new: dropDatabase invokes callback with EDBMISSING error if database doesn't exist

# 0.4.0

 * `listDatabases()` method output was wrong (#5, fixed by @dethtron5000)
 * node.js minimum version is now 0.10 due to asker's `contimer` dependency.

# 0.3.2

 * `update()` method failed to update docs with URL in their _id (#7)
 * update dependencies: asker and nodeunit.

# 0.3.1

 * `node-couchdb` constructor accepts 4th parameter - default timeout for all requests, which is 5 seconds by default.

# 0.3.0

 * Use 5 seconds default timeout for all requests

# 0.2.2

 * Do not use `_serialize` property: we can't guarantee that the cached value won't change later if `_serialize` if set to `false` and value is cached in memory. Now every cached value is serialized by default;
 * Exclude examples and tests from being published to NPM.

# 0.2.1

 * Use [asker](http://browsenpm.org/package/asker) instead of [request](http://browsenpm.org/package/request);
 * Add additional tests;
 * Add .editorconfig file.


src/node-couchdb.js
'use strict';
import crypto from 'crypto';
import AbortController from 'abort-controller';

// @see https://wiki.apache.org/couchdb/HTTP_view_API#Querying_Options
// @see https://github.com/1999/node-couchdb/issues/9
const KEYS_TO_ENCODE = ['key', 'keys', 'startkey', 'endkey'];

class RequestError extends Error {
    constructor(code, message, body) {
        super(message);

        this.code = code;
        this.body = body;
    }
}

export default class NodeCouchDB {
    constructor(opts = {}) {
        const instanceOpts = Object.assign({
            protocol: 'http',
            host: '127.0.0.1',
            port: 5984,
            cache: null,
            timeout: 5000,
            auth: null
        }, opts);

        this._baseUrl = `${instanceOpts.protocol}://`;
        this._baseUrl += `${instanceOpts.host}:${instanceOpts.port}`;

        this._cache = instanceOpts.cache;
        this._timeoutMs = instanceOpts.timeout;

        const defaultHeaders = {
            'user-agent': 'node-couchdb/1',
            'content-type': 'application/json'
        };

        if (instanceOpts.auth) {
            const str = `${instanceOpts.auth.user}:${instanceOpts.auth.pass}`;
            const b64 = Buffer.from(str, 'utf8').toString('base64');
            defaultHeaders['authorization'] = 'Basic ' + b64;
        }


        this._fetchDefaultOpts = {
            headers: defaultHeaders,
        };
    }

    /**
     * Use new cache mechanism (an "invalidate" method of the old cache machanism will be invoked)
     * This method is useful if you want to call GC manually
     *
     * @param {Object} cacheObj
     * @return {Undefined}
     */
    useCache(cache) {
        try {
            this._cache.invalidate();
        } catch (e) {}

        this._cache = cache;
    }

    /**
     * Get the list of all databases. Returns a promise which is
     * - resolved with {Array} list of databases
     * - rejected with `request` original error
     *
     * @return {Promise}
     */
    listDatabases() {
        return this._fetchWrapped(`${this._baseUrl}/_all_dbs`).then(({body}) => body);
    }

    /**
     * Creates a database. Returns a promise which is
     * - resolved with no arguments
     * - rejected with `request` original error
     *
     * @param {String} dbName
     * @return {Promise}
     */
    createDatabase(dbName) {
        return this._fetchWrapped(`${this._baseUrl}/${dbName}`, {
            method: 'PUT',
        }).then(({res, body}) => {
            // database already exists
            if (res.status === 412) {
                throw new RequestError('EDBEXISTS', `Database already exists: ${dbName}`, body);
            }

            if (res.status === 401) {
                throw new RequestError('ENOTADMIN', `Should be authorized as admin to create database: ${res.status}`, body);
            }

            if (res.status === 400) {
                throw new RequestError('EBADREQUEST', res.body.reason, body);
            }

            if (res.status !== 201 && res.status !== 202) {
                throw new RequestError('EUNKNOWN', `Unexpected status code while creating database ${dbName}: ${res.status}`, body);
            }
        });
    }

    /**
     * Drops database by its name. Returns a promise which is
     * - resolved with no arguments
     * - rejected with `request` original error
     *
     * @param {String} dbName
     * @return {Promise}
     */
    dropDatabase(dbName) {
        const url = `${this._baseUrl}/${dbName}/`;
        return this._fetchWrapped(url, {
            method: 'DELETE',
        }).then(({res, body}) => {
            // database not found
            if (res.status === 404) {
                throw new RequestError('EDBMISSING', `Database not found: ${dbName}`, body);
            }

            if (res.status === 401) {
                throw new RequestError('ENOTADMIN', `Should be authorized as admin to delete database: ${res.status}`, body);
            }

            if (res.status !== 200 && res.status !== 202) {
                throw new RequestError('EUNKNOWN', `Unexpected status code while deleting database ${dbName}: ${res.status}`, body);
            }
        });
    }

    /**
     * Fetch data from CouchDB. Returns a promise which is
     * - resolved with {data, headers, status} object
     * - rejected with `request` original error
     *
     * @param {String} dbName database name
     * @param {String} uri document ID or design view
     * @param {Object} [query] query options as key: value
     * @return {Promise}
     */
    get(dbName, uri, query = {}) {
        const url = new URL(`${this._baseUrl}/${dbName}/${uri}`); 

        for (let prop in query) {
            if (KEYS_TO_ENCODE.includes(prop)) {
                url.searchParams.set(prop, JSON.stringify(query[prop]));
            } else {
                url.searchParams.set(prop, query[prop]);
            }
        }

        return this._fetchWrapped(url).then(({res, body}) => {
            if (res.status === 404) {
                throw new RequestError('EDOCMISSING', 'Document is not found', body);
            }

            if (res.status !== 200 && res.status !== 304) {
                throw new RequestError('EUNKNOWN', `Unexpected status code while fetching documents from the database: ${res.status}`, body);
            }

            if (res.status === 200 && this._cache) {
                const cacheKey = this._getCacheKey(url);

                this._cache.set(cacheKey, {
                    body,
                    etag: res.headers.get('ETag')
                });
            }

            return {
                data: body,
                headers: res.headers,
                status: res.status
            };
        });
    }

    /**
     * Fetch attachment from CouchDB. Returns a promise which is
     * - resolved with {data, headers, status} object
     * - rejected with `request` original error
     *
     * @param {String} dbName database name
     * @param {String} docId document id
     * @param {String} attachmentName attachment name
     * @param {String} docRevision document revision
     * @return {Promise}
     */
    getAttachment(dbName, docId, attachmentName, docRevision) {

        const url = new URL(`${this._baseUrl}/${dbName}/${docId}/${attachmentName}`);
        url.searchParams.set('rev', docRevision);

        return this._fetchWrapped(url).then(({res, body}) => {
            if (res.status === 404) {
                throw new RequestError('EDOCMISSING', 'Attachment is not found', body);
            }

            if (res.status !== 200 && res.status !== 304) {
                throw new RequestError('EUNKNOWN', `Unexpected status code while fetching attachment from the database: ${res.status}`, body);
            }

            if (res.status === 200 && this._cache) {
                const cacheKey = this._getCacheKey(url);

                this._cache.set(cacheKey, {
                    body,
                    etag: res.headers.get('ETag')
                });
            }

            return {
                data: body,
                headers: res.headers,
                status: res.status
            };
        });
    }

    /**
     * Insert document into CouchDB. Returns a promise which is
     * - resolved with {data, headers, status} object
     * - rejected with `request` original error
     *
     * @param {String} dbName database name
     * @param {Object} data
     * @return {Promise}
     */
    insert(dbName, data) {
        const url = `${this._baseUrl}/${dbName}`;
        return this._fetchWrapped(url, {
            method: 'POST',
            body: JSON.stringify(data)
        }).then(({res, body}) => {
            this._checkDocumentManipulationStatus(res.status, body)

            if (res.status !== 201 && res.status !== 202) {
                throw new RequestError('EUNKNOWN', `Unexpected status code while inserting document into the database: ${res.status}`, body);
            }

            return {
                data: body,
                headers: res.headers,
                status: res.status
            };
        });
    }

    /**
     * Insert document into CouchDB. Returns a promise which is
     * - resolved with {data, headers, status} object
     * - rejected with `request` original error
     *
     * @param {String} dbName database name
     * @param {String} docId document id
     * @param {String} attachmentName attachment name
     * @param {String} body attachment body
     * @param {String} docRevision document revision
     * @return {Promise}
     */
    insertAttachment(dbName, docId, attachmentName, body, docRevision) {
        const url = new URL(`${this._baseUrl}/${dbName}/${encodeURIComponent(docId)}/attachment`);
        url.searchParams.set('rev', docRevision);
        return this._fetchWrapped(url, {
            method: 'PUT',
            body: JSON.stringify(body)
        }).then(({res, body}) => {
            if (res.status === 409) {
                throw new RequestError('EDOCCONFLICT', 'Document insert conflict - Document’s revision wasn’t specified or it’s not the latest', body);
            }

            return {
                data: body,
                headers: res.headers,
                status: res.status
            };
        });
    }

    /**
     * Update a document in CouchDB. Returns a promise which is
     * - resolved with {data, headers, status} object
     * - rejected with `request` original error
     *
     * @param {String} dbName database name
     * @param {Object} data should contain both "_id" and "_rev" fields
     * @return {Promise}
     */
    update(dbName, data) {
        if (!data._id || !data._rev) {
            const err = new Error('Both _id and _rev fields should exist when updating the document');
            err.code = 'EFIELDMISSING';

            return Promise.reject(err);
        }

        const url = `${this._baseUrl}/${dbName}/${encodeURIComponent(data._id)}`;

        return this._fetchWrapped(url, {
            method: 'PUT',
            body: JSON.stringify(data)
        }).then(({res, body}) => {
            this._checkDocumentManipulationStatus(res.status, body)


            if (!(res.status >= 200 && res.status <= 202)) {
                throw new RequestError('EUNKNOWN', `Unexpected status code while inserting document into the database: ${res.status}`, body);
            }

            return {
                data: body,
                headers: res.headers,
                status: res.status
            };
        });
    }

    /**
     * Delete a document in the database. Returns a promise which is
     * - resolved with {data, headers, status} object
     * - rejected with `request` original error
     *
     * @param {String} dbName database name
     * @param {String} docId document id
     * @param {String} docRevision document revision
     * @return {Promise}
     */
    del(dbName, docId, docRevision) {
        const url = new URL(`${this._baseUrl}/${dbName}/${encodeURIComponent(docId)}`);
        url.searchParams.set('rev', docRevision);
        return this._fetchWrapped(url, {
            method: 'DELETE',
        }).then(({res, body}) => {
            this._checkDocumentManipulationStatus(res.status, body)

            if (res.status !== 200) {
                throw new RequestError('EUNKNOWN', `Unexpected status code while deleting document: ${res.status}`, body);
            }

            return {
                data: body,
                headers: res.headers,
                status: res.status
            };
        });
    }

    /**
     * Fetch data from CouchDB using Mango API. Returns a promise which is
     * - resolved with {data, headers, status} object
     * - rejected with `request` original error
     *
     * @param {String} dbName database name
     * @param {String|Object} mangoQuery Mango query as json string or javascript object
     * @param {Object} [query] query options as key: value
     * @return {Promise}
     */
    mango(dbName, mangoQuery) {
        const url = new URL(`${this._baseUrl}/${dbName}/_find`);

        if (typeof mangoQuery === 'string') {
            try {
                mangoQuery = JSON.parse(mangoQuery);
            } catch (e) {
                return Promise.reject(new RequestError('EBADREQUEST', 'The Mango query parameter is not parsable.'));
            }
        }

        if (typeof mangoQuery !== 'object') {
            return Promise.reject(new RequestError('EBADREQUEST', 'Invalid Mango query parameter.'));
        }

        const requestOpts = {
            method: 'POST',
            body: JSON.stringify(mangoQuery),
        };

        return this._fetchWrapped(url, requestOpts).then(({res, body}) => {
            this._checkServerVersion(res.headers.get('Server'), 2);

            if (res.status === 404) {
                throw new RequestError('EDOCMISSING', 'Document is not found', body);
            }

            if (res.status !== 200 && res.status !== 304) {
                throw new RequestError('EUNKNOWN', `Unexpected status code while fetching documents from the database: ${res.status}`, body);
            }

            return {
                data: body,
                headers: res.headers,
                status: res.status
            };
        });
    }

    /**
     * Delete a document in the database. Returns a promise which is
     * - resolved with {data, headers, status} object
     * - rejected with `request` original error
     *
     * @param {String} dbName database name
     * @param {String} docId document id
     * @param {String} attachmentName attachment name
     * @param {String} docRevision document revision
     * @return {Promise}
     */
    delAttachment(dbName, docId, attachmentName, docRevision) {
        const url = new URL(`${this._baseUrl}/${dbName}/${encodeURIComponent(docId)}/${encodeURIComponent(attachmentName)}`);
        url.searchParams.set('rev', docRevision);

        return this._fetchWrapped(url, {
            method: 'DELETE',
        }).then(({res, body}) => {
            if (res.status === 404) {
                throw new RequestError('EDOCMISSING', 'Attachment is not found', body);
            }

            if (res.status !== 200) {
                throw new RequestError('EUNKNOWN', `Unexpected status code while deleting attachment: ${res.status}`, body);
            }

            return {
                data: body,
                headers: res.headers,
                status: res.status
            };
        });
    }

    /**
     * Calls an update function in the database. Returns a promise which is
     * - resolved with {data, headers, status} object
     * - rejected with `request` original error
     *
     * @param  {String} dbName             database name
     * @param  {String} designDocument     design document name
     * @param  {String} updateFunctionName update function name
     * @param  {Object} queryString        query string parameters
     * @param  {String} docId              document id

     * @return {Promise}
     */
    updateFunction(dbName, designDocument, updateFunctionName, queryString, docId ) {
        const method = docId ? 'PUT' : 'POST';
        queryString = queryString || {};

        let url;

        if (method === 'PUT') {
            url = new URL(`${this._baseUrl}/${dbName}/_design/${designDocument}/_update/${updateFunctionName}/${docId}`);
        } else {
            url = new URL(`${this._baseUrl}/${dbName}/_design/${designDocument}/_update/${updateFunctionName}`);
        }
        
        for (let prop in queryString) {
            url.searchParams.set(prop, queryString[prop]);
        }

        return this._fetchWrapped(url, {
            method: method,
        }).then(({res, body}) => {
            if (res.status === 404) {
                throw new RequestError('EDOCMISSING', 'Design document is not found', body);
            }

            if (res.status !== 200 && res.status !== 201 && res.status !== 202) {
                throw new RequestError('EUNKNOWN', `Unexpected status code while calling update function: ${res.status}`, body);
            }

            return {
                data: body,
                headers: res.headers,
                status: res.status
            };
        });
    }

    /**
     * Get UUIDs for new documents. Returns a promise which is
     * - resolved with array of new unique ids
     * - rejected with `request` original error
     *
     * @param {Number} [count = 1] number of IDs you want to get
     * @return {Promise}
     */
    uniqid(count = 1) {
        const url = new URL(`${this._baseUrl}/_uuids`); 
        url.searchParams.set('count', count);

        return this._fetchWrapped(url).then(({body}) => {
            return body.uuids;
        });
    }

    /**
     * Check the status code of a documentation manipulation like INSERT, UPDATE, DELETE
     *
     * @param {Number} statusCode
     * @param {Object} body
     * @throws {RequestError}
     */
    _checkDocumentManipulationStatus(statusCode, body) {
      if (statusCode === 400) {
          throw new RequestError('EBADREQUEST', 'Invalid request body or parameters', body);
      }

      if (statusCode === 401) {
          throw new RequestError('EUNAUTHORIZED', 'Write privileges required', body);
      }

      if (statusCode === 404) {
          throw new RequestError('EDOCMISSING', 'Document not found', body);
      }

      if (statusCode === 409) {
          throw new RequestError('EDOCCONFLICT', 'Document insert conflict', body);
      }
    }

    /**
     * Requests wrapper. Checks for cache first for GET requests.
     * Should be invoked with arguments suitable for `request`
     *
     * @return {Promise}
     */
    async _fetchWrapped(url, opts) {

        opts = opts || {};

        const cacheKey = this._getCacheKey(url);
        const whenCacheChecked = (!this._cache || (opts.method && opts.method !== 'GET'))
            ? Promise.resolve({})
            : this._cache.get(cacheKey);

        return whenCacheChecked.then(cache => {

            // cache plugin returns null if record doesn't exist
            const {
                etag,
                body: cacheBody
            } = cache || {};

            return new Promise(async (resolve, reject) => {
                opts = Object.assign({ ... this._fetchDefaultOpts }, opts);

                if (etag) {
                    opts.headers = opts.headers || {};
                    opts.headers['if-none-match'] = etag;
                }

                const controller = new AbortController();
                const timeout = setTimeout(() => controller.abort(), this._timeoutMs);
                opts.signal = controller.signal;

                try {
                    const res = await fetch(url, opts);

                    let data = null;
                    if (res.ok) {
                        const contentType = res.headers.get('content-type');

                        if (contentType.includes('application/json')) {
                            data = await res.json();
                        } else if (contentType.includes('text/html')){
                            data = await res.text();
                        }
                    }

                    resolve({
                        res,
                        body: data || cacheBody
                    });
                } catch (err) {
                    reject(err);
                } finally {
                    clearTimeout(timeout);
                }
            });
        });
    }


    /**
     * Gets cache key built from request options
     *
     * @param {Object} requestOpts
     * @return {String}
     */
    _getCacheKey(url) {
        const cacheKeyFull = url.toString();
        return crypto.createHash('md5').update(cacheKeyFull).digest('hex');
    }

    /**
     * @param {String} serverHeader
     * @param {Number} minServerVersion
     */
    _checkServerVersion(serverHeader, minServerVersion = 1) {
        const serverVersion = serverHeader.match(/^CouchDB\/([\d]+)/);

        if (!serverVersion || !serverVersion[1]) {
            throw new RequestError('ESERVERNOTSUPPORTED', `Server is not supported: ${serverHeader}`);
        }

        if (serverVersion[1] < minServerVersion) {
            throw new RequestError('ESERVEROLD', `Server version is too old for using this API: ${minServerVersion} (expected), ${serverHeader} (actual)`);
        }
    }
};
